match doesn't work either in return statement or maps

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** master
**OS:** atch

**What did you do?**
```
	return {
		'token': cfg['token'][0]
		'sleep': match cfg['sleep'][0].len {
			0 { '60' }
			else { cfg['sleep'][0] }
		}
		'every': match cfg['every'][0].len {
			0 { '5' }
			else { cfg['every'][0] }
		}
	}

```

Error

```
/tmp/v/gh_notify.7498108735114691329.tmp.c: In function ‘main__config’:
/tmp/v/gh_notify.7498108735114691329.tmp.c:17969:388: error: expected expression before ‘int’
17969 |  return new_map_init_2(&map_hash_string, &map_eq_string, &map_clone_string, &map_free_string, 3, sizeof(string), sizeof(string), _MOV((string[3]){_SLIT("token"), _SLIT("sleep"), _SLIT("every"), }), _MOV((string[3]){(*(string*)/*ee elem_typ */array_get((*(array_string*)map_get_1(ADDR(map, cfg), &(string[]){_SLIT("token")}, &(array_string[]){ __new_array(0, 1, sizeof(string)) })), 0)), int _t812 = (*(string*)/*ee elem_typ */array_get((*(array_string*)map_get_1(ADDR(map, cfg), &(string[]){_SLIT("sleep")}, &(array_string[]){ __new_array(0, 1, sizeof(string)) })), 0)).len;
   
builder error:
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

```

sorry about that @serkonda7 

Here you go https://gist.github.com/tami5/4c546eaa529c924d02f6df3ce4f1b998#file-gh_notify-v-L57

vendor config can be found here https://github.com/gslicer/configparser/blob/master/configparser/cp.v


```
/tmp/v/gh_notify.vc.11477052081823722913.tmp.c: In function ‘main__config’:
/tmp/v/gh_notify.vc.11477052081823722913.tmp.c:17947:388: error: expected expression before ‘int’
17947 |  return new_map_init_2(&map_hash_string, &map_eq_string, &map_clone_string, &map_free_string, 3, sizeof(string), sizeof(string), _MOV((string[3]){_SLIT("token"), _SLIT("sleep"), _SLIT("every"), }), _MOV((string[3]){(*(string*)/*ee elem_typ */array_get((*(array_string*)map_get_1(ADDR(map, cfg), &(string[]){_SLIT("token")}, &(array_string[]){ __new_array(0, 1, sizeof(string)) })), 0)), int _t812 = (*(string*)/*ee elem_typ */array_get((*(array_string*)map_get_1(ADDR(map, cfg), &(string[]){_SLIT("sleep")}, &(array_string[]){ __new_array(0, 1, sizeof(string)) })), 0)).len;
      |                                                                                                                                                                                                                                                                                                                                                                                                    ^~~
...


```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5416,6 +5416,7 @@ pub fn (mut c Checker) map_init(mut node ast.MapInit) table.Type {
 		}
 		val := node.vals[i]
 		key_type := c.expr(key)
+		c.expected_type = val0_type
 		val_type := c.expr(val)
 		if !c.check_types(key_type, key0_type) {
 			msg := c.expected_msg(key_type, key0_type)

--- a/vlib/v/tests/match_in_map_init_test.v
+++ b/vlib/v/tests/match_in_map_init_test.v
@@ -1,0 +1,24 @@
+fn test_match_in_map_init() {
+	ret := foo()
+	println(ret)
+	assert ret == {'token': 'a', 'sleep': '30', 'every': '1'}
+}
+
+fn foo() map[string]string {
+	mut cfg := map[string][]string{}
+	cfg['token'] = ['a', 'b']
+	cfg['sleep'] = ['30', '60']
+	cfg['every'] = ['1', '5']
+
+	return {
+		'token': cfg['token'][0]
+		'sleep': match cfg['sleep'][0].len {
+			0 { '60' }
+			else { cfg['sleep'][0] }
+		}
+		'every': match cfg['every'][0].len {
+			0 { '5' }
+			else { cfg['every'][0] }
+		}
+	}
+}

--- a/vlib/v/tests/match_in_map_init_test.v
+++ b/vlib/v/tests/match_in_map_init_test.v
@@ -1,7 +1,7 @@
 fn test_match_in_map_init() {
 	ret := foo()
 	println(ret)
-	assert ret == {'token': 'a', 'sleep': '30', 'every': '1'}
+	assert ret == map{'token': 'a', 'sleep': '30', 'every': '1'}
 }
 
 fn foo() map[string]string {
@@ -10,7 +10,7 @@ fn foo() map[string]string {
 	cfg['sleep'] = ['30', '60']
 	cfg['every'] = ['1', '5']
 
-	return {
+	return map{
 		'token': cfg['token'][0]
 		'sleep': match cfg['sleep'][0].len {
 			0 { '60' }

